docs/portingguide/src/howto/nogc.md

```diff
@@ -1,50 +1,283 @@
+
+We describe the most common metadata specifications and potential metadata locations:
+
+  1. **Unlog bit metadata:** A global 1-bit metadata used by generational collectors to keep track of cross-generational pointers. Usually side metadata.
+  2. **Forwarding bits and pointer**: A local metadata used by copying policies to store forwarding bits (2-bits) and forwarding pointers (word size). Often runtimes require word-aligned addresses which means we can use the last two bits in the object header (due to alignment) and the entire object header to store the forwarding bits and pointer respectively. Almost always in the header.
+  3. **Mark bits**: A local 1-bit metadata used by tracing collectors to mark seen objects. Like with the forwarding bits, we can often steal the last bit in the object header for the mark bit. Though some bindings such as the OpenJDK binding prefer to have the mark bits in side metadata to allow for bulk updates.
+  4. **Large object space mark and nursery bits**: A local 2-bit metadata used by the the large object space to mark objects and set objects as "newly allocated". This should always be in the header. For large objects, we can add an extra word to store this metadata since the metadata size is insignificant in comparison to the object size. See [here](https://github.com/mmtk/mmtk-core/issues/847) for more information.
+
+#### `ObjectReference` vs `Address`
+
+A key principle in MMTk is the distinction between [`ObjectReference`](https://www.mmtk.io/mmtk-core/public-doc/util/address/struct.ObjectReference.html) and [`Address`](https://www.mmtk.io/mmtk-core/public-doc/util/address/struct.Address.html). The idea is that very few operations are allowed on an `ObjectReference`. For example, MMTk does not allow address arithmetic on `ObjectReference`s. This allows us to preserve memory-safety, only performing unsafe operations when required, and gives us a cleaner and more flexible abstraction to work with as it can allow object handles or offsets etc. `Address`, on the other hand, represents an arbitrary machine address.
+
+You might be interested in reading the *Demystifying Magic: High-level Low-level Programming* paper[^3] which describes the above in more detail.
+
+[^3]: https://users.cecs.anu.edu.au/~steveb/pubs/papers/vmmagic-vee-2009.pdf
+
+**Note:** Currently the `ObjectReference` is not fully opaque to MMTk, so there is some abstraction leakage, but [work is underway](https://github.com/mmtk/mmtk-core/issues/686) to make `ObjectReference` fully opaque to MMTk.
+
+You must understand the difference between an `ObjectReference` and an `Address` in your runtime and implement the conversion functions accordingly. It is often the case that there is no difference between the two, but you must make sure it is the case.
+
+#### Miscellaneous configuration options
```

Comment:
Well it needs to be sign posted somewhere. I think it makes sense to at least mention them in the porting guide. I am not describing what each constant does, only 1-2 constants.

Reply:
See [my comments above](https://github.com/mmtk/mmtk-core/pull/857#discussion_r1239263078). We can always document/comment the code properly, and use a link here to refer to the code. In my opinion, this porting guide is not some sort of supplementary comments/docs for the code -- it is guiding people through the code with the focus of implementing a binding. If we lack of sufficient comments in the code, we should add comments to the code, not here.

But anyway, I am relaxed on this. If you would like to list things in the guide, and more clearly describe what they are here, I am okay with it.

Reply:
Sign posted somewhere in the sense that porters may not even be aware of the existence of the constants. They should be mentioned somewhere, preferably in the porting guide, so that they can go and edit it. I agree that the guide may get stale which is why I think it might be better to just mention one constant and then say "Hey you should go and check these other constants out and you may have to change them depending on your runtime's requirements".